includes/abilities-api/class-wp-abilities-category-registry.php

```diff
@@ -0,0 +1,240 @@
+				__METHOD__,
+				esc_html__( 'Category slug must contain only lowercase alphanumeric characters and dashes.' ),
+				'0.3.0'
+			);
+			return null;
+		}
+
+		/**
+		 * Filters the category arguments before they are validated and used to instantiate the category.
+		 *
+		 * @since 0.3.0
+		 *
+		 * @param array<string,mixed> $args The arguments used to instantiate the category.
+		 * @param string              $slug The slug of the category.
+		 */
+		$args = apply_filters( 'register_ability_category_args', $args, $slug );
+
+		try {
+			// WP_Ability_Category::prepare_properties() will throw an exception if the properties are invalid.
+			$category = new WP_Ability_Category( $slug, $args );
+		} catch ( \InvalidArgumentException $e ) {
+			_doing_it_wrong(
+				__METHOD__,
+				esc_html( $e->getMessage() ),
+				'0.3.0'
+			);
+			return null;
+		}
+
+		$this->registered_categories[ $slug ] = $category;
+		return $category;
+	}
+
+	/**
+	 * Unregisters a category.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_unregister_ability_category()` function.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @see wp_unregister_ability_category()
+	 *
+	 * @param string $slug The slug of the registered category.
+	 * @return ?\WP_Ability_Category The unregistered category instance on success, null on failure.
+	 */
+	public function unregister( string $slug ): ?WP_Ability_Category {
+		if ( ! $this->is_registered( $slug ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Category slug. */
+				sprintf( esc_html__( 'Category "%s" not found.' ), esc_attr( $slug ) ),
+				'0.3.0'
+			);
+			return null;
+		}
+
+		$unregistered_category = $this->registered_categories[ $slug ];
+		unset( $this->registered_categories[ $slug ] );
+
+		return $unregistered_category;
+	}
+
+	/**
+	 * Retrieves the list of all registered categories.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_get_ability_categories()` function.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @see wp_get_ability_categories()
+	 *
+	 * @return array<string,\WP_Ability_Category> The array of registered categories.
+	 */
+	public function get_all_registered(): array {
+		return $this->registered_categories;
+	}
+
+	/**
+	 * Checks if a category is registered.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param string $slug The slug of the category.
+	 * @return bool True if the category is registered, false otherwise.
+	 */
+	public function is_registered( string $slug ): bool {
+		return isset( $this->registered_categories[ $slug ] );
+	}
+
+	/**
+	 * Retrieves a registered category.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_get_ability_category()` function.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @see wp_get_ability_category()
+	 *
+	 * @param string $slug The slug of the registered category.
+	 * @return ?\WP_Ability_Category The registered category instance, or null if it is not registered.
+	 */
+	public function get_registered( string $slug ): ?WP_Ability_Category {
```

Comment:
This is a registry, so everything is registered, so what we're really getting here is a category.
```suggestion
	public function get_category( string $slug ): ?WP_Ability_Category {
```

Reply:
I think it's fine as is because these names containing `registered` keyword are used in several places in WP core. 3rd party consumers won't use these names but function helpers instead.

Reply:
Example: 

https://developer.wordpress.org/reference/classes/wp_block_type_registry/get_registered/

Reply:
Slightly more relevant example for posterity (WP_Block_Type IRL more mirrors our WP_Ability): https://github.com/WordPress/wordpress-develop/blob/6.8.3/src/wp-includes/class-wp-block-pattern-categories-registry.php

Reply:
For consistency sake I'll agree. For readability, I'll kindly protest that `$registry->get_registered()` tells me nothing about what I'm getting. 😁